From c5ea7a7458d74dd1aebceab6d4a5a6f533c6f8ba Mon Sep 17 00:00:00 2001
Date: Mon, 29 Jan 2024 01:48:56 +0100
Subject: [PATCH 2/3] Update CMakeLists with iOS implementation

---
 CMakeLists.txt | 35 +++++++++++++++++++++++++----------
 1 file changed, 25 insertions(+), 10 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3d81062..218ba7c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -267,14 +267,25 @@ elseif(UNIX)
 
   if(APPLE)
     set(CMAKE_MACOSX_RPATH 1)
-    target_sources(portaudio PRIVATE
-      src/hostapi/coreaudio/pa_mac_core.c
-      src/hostapi/coreaudio/pa_mac_core_blocking.c
-      src/hostapi/coreaudio/pa_mac_core_blocking.h
-      src/hostapi/coreaudio/pa_mac_core_internal.h
-      src/hostapi/coreaudio/pa_mac_core_utilities.c
-      src/hostapi/coreaudio/pa_mac_core_utilities.h
-    )
+    if(IOS)
+      target_sources(portaudio PRIVATE
+        src/hostapi/coreaudio_ios/pa_ios_core.c
+        src/hostapi/coreaudio_ios/pa_ios_core_blocking.c
+        src/hostapi/coreaudio_ios/pa_ios_core_blocking.h
+        src/hostapi/coreaudio_ios/pa_ios_core_internal.h
+        src/hostapi/coreaudio_ios/pa_ios_core_utilities.c
+        src/hostapi/coreaudio_ios/pa_ios_core_utilities.h
+      )
+    else()
+      target_sources(portaudio PRIVATE
+        src/hostapi/coreaudio/pa_mac_core.c
+        src/hostapi/coreaudio/pa_mac_core_blocking.c
+        src/hostapi/coreaudio/pa_mac_core_blocking.h
+        src/hostapi/coreaudio/pa_mac_core_internal.h
+        src/hostapi/coreaudio/pa_mac_core_utilities.c
+        src/hostapi/coreaudio/pa_mac_core_utilities.h
+      )
+    endif()
     target_include_directories(portaudio PRIVATE src/hostapi/coreaudio)
     set(PORTAUDIO_PUBLIC_HEADERS "${PORTAUDIO_PUBLIC_HEADERS}" include/pa_mac_core.h)
 
@@ -290,8 +301,12 @@ elseif(UNIX)
         -Wl,-framework,CoreFoundation
         -Wl,-framework,CoreServices
     )
-    target_compile_definitions(portaudio PUBLIC PA_USE_COREAUDIO=1)
-    set(PKGCONFIG_CFLAGS "${PKGCONFIG_CFLAGS} -DPA_USE_COREAUDIO=1")
+    if(IOS)
+      target_compile_definitions(portaudio PUBLIC PA_USE_COREAUDIO_IOS=1)
+    else()
+      target_compile_definitions(portaudio PUBLIC PA_USE_COREAUDIO=1)
+      set(PKGCONFIG_CFLAGS "${PKGCONFIG_CFLAGS} -DPA_USE_COREAUDIO=1")
+    endif()
 
     # Use C11 so that we can make use of atomic library and avoid deprecation errors.
     set_property(TARGET portaudio PROPERTY C_STANDARD 11)
-- 
2.43.0

